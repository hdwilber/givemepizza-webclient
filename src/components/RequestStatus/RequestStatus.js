import React from 'react'
import PropTypes from 'prop-types'
import styles from './styles.module.css'

export const messages = {
  done: 'Request successful',
  wrong: 'Something went wrong',
  loading: 'Loading request',
  notStarted: 'Request not started',
}
function parseStatus(status) {
  if (status) {
    const { loading, loaded, error } = status
    if (!loading && loaded) {
      if (error) {
        return messages.wrong
      }
      return messages.done
    }
    if (loading) {
      return messages.loading
    }
  }
  return messages.notStarted
}
class RequestStatus extends React.PureComponent {
  render() {
    const { requestsList } = this.props

    return (
      <div className={styles.status}>
        { requestsList.map(request => {
            const { name, status } = request
            return (
              <p key={name}>
                <b>{name}</b>&nbsp;
                { parseStatus(status) }
              </p>
            )
          })
        }
      </div>
    )
  }
}

RequestStatus.propTypes = {
  requestsList: PropTypes.array,
}
RequestStatus.defaultProps = {
  requestsList: [],
}

export default RequestStatus
