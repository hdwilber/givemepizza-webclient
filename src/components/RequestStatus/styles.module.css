.status {
	position: fixed;
	bottom: 0;
	width: 100%;
  max-height: 250px;
  overflow-y: scroll;
}
