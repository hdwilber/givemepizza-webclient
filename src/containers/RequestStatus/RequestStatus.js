import React from 'react'
import RequestStatusComponent from '../../components/RequestStatus'

class RequestStatus extends React.PureComponent {
  render() {
    const { show, list } = this.props
    if(show) {
      return (
        <RequestStatusComponent requestsList={list} />
      )
    }
    return null
  }
}

export default RequestStatus
