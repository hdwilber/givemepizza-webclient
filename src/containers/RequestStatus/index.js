import { connect } from 'react-redux'
import _get from 'lodash/get'
import RequestStatus from './RequestStatus'

function mapStateToProps(state) {
  const { status: { show, requests } } = state
  return {
    list: requests.map(info => {
      const { path, name, description } = info
      return {
        status: _get(state, path, null),
        name,
        description,
      }
    }),
    show,
  }
}

function mapDispatchToProps(dispatch) {
  return {}
}

export default connect(mapStateToProps, mapDispatchToProps) (RequestStatus)
