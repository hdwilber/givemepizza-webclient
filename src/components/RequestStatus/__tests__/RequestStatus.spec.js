import React from 'react'
import { mount } from 'enzyme'
import RequestStatus, { messages } from '../RequestStatus'

function setupProps() {
  return {
    requestsList: [
      {
        name: 'Some loading request',
        status: {
          loading: true,
          loaded: false,
        }
      },
      {
        name: 'Some correctly loaded request',
        status: {
          loading: false,
          loaded: true,
          data: {},
        }
      },
      {
        name: 'Some failing request',
        status: {
          loading: false,
          loaded: true,
          error: {
          }
        }
      },
      {
        name: 'Not loaded request',
        status: null,
      },
    ]
  }
}

function setup(props) {
  return mount(<RequestStatus {...props} />)
}

describe('<RequestStatus />', () => {
  let props
  let wrapper
  beforeEach(() => {
    props = setupProps()
    wrapper = setup(props)
  })

  it('should match snapshot', () => {
    expect(<RequestStatus {...props} />).toMatchSnapshot()
  })

  it('should render correctly the request information according to their status', () => {
    const requestsWrapper = wrapper.find('p')
    const texts = [messages.loading, messages.done, messages.wrong, messages.notStarted]
    texts.forEach((text, index) => {
      expect(requestsWrapper.at(index).text()).toMatch(text)
    })
  })
})
