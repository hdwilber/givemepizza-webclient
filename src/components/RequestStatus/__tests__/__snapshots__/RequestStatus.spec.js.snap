// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestStatus /> should match snapshot 1`] = `
<RequestStatus
  requestsList={
    Array [
      Object {
        "name": "Some loading request",
        "status": Object {
          "loaded": false,
          "loading": true,
        },
      },
      Object {
        "name": "Some correctly loaded request",
        "status": Object {
          "data": Object {},
          "loaded": true,
          "loading": false,
        },
      },
      Object {
        "name": "Some failing request",
        "status": Object {
          "error": Object {},
          "loaded": true,
          "loading": false,
        },
      },
      Object {
        "name": "Not loaded request",
        "status": null,
      },
    ]
  }
/>
`;
